### PR TITLE
Python Bindings for launching PySpark Jobs from the JVM 

### PR DESCRIPTION
**Pull request by ifilonenko**
## What changes were proposed in this pull request?

The changes that were proposed in the pull request are the following:
1. Using separate docker images, built on-top of spark-base, for PySpark jobs.
These images differ with the inclusion of python and pyspark specific environment variables. The user-entry point also differs for driver-py as you must include the location of the primary PySpark file and distributed py-files in addition to driver args. 
2. New FileMountingTrait that is generic enough for both SparkR and PySpark to handle passing in the proper arguments for PythonRunner and RRunner. This FileMounter uses the filesDownloadPath resolved in the DriverInitComponent to ensure that file paths are correct. These file-paths are stored as environmental variables that are mounted on the driver pod. 
3. Inclusion of integration tests for PySpark (TODO: Build environment identical to distribution python environment to run the tests) 
4. Match statements to account for varying arguments and malformed inputs which may include nulls or a mix of local:// and file:// file-types. 

## Example Spark Submit
This is an example spark-submit that uses the custom pyspark docker images and distributes the staged `sort.py` file across the cluster. The entry point for the driver is:
`org.apache.spark.deploy.PythonRunner <FILE_DOWNLOADS_PATH>/pi.py <FILE_DOWNLOADS_PATH>/sort.py 100`

```bash
bin/spark-submit \
  --deploy-mode cluster \
  --master k8s://<k8s-api-url> \
  --kubernetes-namespace default \
  --conf spark.executor.memory=500m \
  --conf spark.driver.memory=1G \
  --conf spark.driver.cores=1 \
  --conf spark.executor.cores=1 \
  --conf spark.executor.instances=1 \
  --conf spark.app.name=spark-pi \
  --conf spark.kubernetes.driver.docker.image=spark-driver-py:latest \
  --conf spark.kubernetes.executor.docker.image=spark-executor-py:latest \
  --conf spark.kubernetes.initcontainer.docker.image=spark-init:latest \
  --conf spark.kubernetes.resourceStagingServer.uri=http://192.168.99.100:31000 \
  --py-files examples/src/main/python/sort.py \
  examples/src/main/python/pi.py 100
```
## How was this patch tested?

This was fully tested by building a make_distribution environment and running on a local minikube cluster with a single executor. The following command is an example submission:

```bash
$ build/mvn install -Pkubernetes -pl resource-managers/kubernetes/core -am -DskipTests
$ build/mvn compile -T 4C -Pkubernetes -pl resource-managers/kubernetes/core -am -DskipTests
$ dev/make-distribution.sh --pip --tgz -Phadoop-2.7 -Pkubernetes
$ tar -xvf spark-2.1.0-k8s-0.2.0-SNAPSHOT-bin-2.7.3.tgz
$ cd spark-2.1.0-k8s-0.2.0-SNAPSHOT-bin-2.7.3
$ minikube start --insecure-registry=localhost:5000 --cpus 8 --disk-size 20g --memory 8000 --kubernetes-version v1.5.3; eval $(minikube docker-env)
$ # Build all docker images using docker build ....
$ # Make sure staging server is up 
$ kubectl cluster-info
Kubernetes master is running at https://192.168.99.100:8443
KubeDNS is running at https://192.168.99.100:8443/api/v1/proxy/namespaces/kube-system/services/kube-dns
kubernetes-dashboard is running at https://192.168.99.100:8443/api/v1/proxy/namespaces/kube-system/services/kubernetes-dashboard
$ docker images
REPOSITORY                                          
spark-integration-test-asset-server                 
spark-init                                           
spark-resource-staging-server                         
spark-shuffle                                      
spark-executor-py                                    
spark-executor                                      
spark-driver-py                                      
spark-driver                                        
spark-base                                         
$ bin/spark-submit \
  --deploy-mode cluster \
  --master k8s://https://192.168.99.100:8443 \
  --kubernetes-namespace default \
  --conf spark.executor.memory=500m \
  --conf spark.driver.memory=1G \
  --conf spark.driver.cores=1 \
  --conf spark.executor.cores=1 \
  --conf spark.executor.instances=1 \
  --conf spark.app.name=spark-pi \
  --conf spark.kubernetes.driver.docker.image=spark-driver-py:latest \
  --conf spark.kubernetes.executor.docker.image=spark-executor-py:latest \
  --conf spark.kubernetes.initcontainer.docker.image=spark-init:latest \
  --conf spark.kubernetes.resourceStagingServer.uri=http://192.168.99.100:31000 \
  --py-files examples/src/main/python/sort.py \
  local:///opt/spark/examples/src/main/python/pi.py 100
```
Integration and Unit tests have been added. 

## Future Versions of this PR
Launching JVM from Python (log issue)
MemoryOverhead testing (OOMKilled errors)

### CODE REVIEW

**Comment by ifilonenko** [comment]
@mccheah PTAL

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 1]
```diff
@@ -621,14 +621,22 @@ object SparkSubmit {
     if (isKubernetesCluster) {
       childMainClass = "org.apache.spark.deploy.kubernetes.submit.Client"
       if (args.isPython) {
+        childArgs += "--py-file"
         childArgs += args.primaryResource
+        childArgs += "--main-class"
         childArgs += "org.apache.spark.deploy.PythonRunner"
+        childArgs += "--other-py-files"
         childArgs += args.pyFiles
```
Comment (tnachen) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: No need to check if it's empty?

**Comment by mccheah** [comment]
This was not supposed to include #363 so please hold off on reviews until we revert that merge.

**Comment by ifilonenko** [comment]
@tnachen PTAL now

**Comment by mccheah** [comment]
I think this is ok to merge given that we'll be redesigning a lot of the architecture in #365. Feature-wise it's more or less what we need and the integration tests plus manual testing have verified correctness. @foxish did you have any last comments or can we merge this soon?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/Client.scala [hunk 2]
```diff
@@ -292,22 +290,34 @@ private[spark] object Client {
...
       mainClass: String,
       appArgs: Array[String]): Unit = {
+    val isPython = mainAppResource.endsWith(".py")
+    val pythonResource: Option[PythonSubmissionResourcesImpl] =
+      if (isPython) {
+        Option(new PythonSubmissionResourcesImpl(mainAppResource, appArgs))
+      } else None
+    // Since you might need jars for SQL UDFs in PySpark
+    def sparkJarFilter() : Seq[String] =
+      pythonResource.map { p => p.sparkJars}.getOrElse(
```
Comment (tnachen) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/Client.scala]: Style: {p => p.SparkJars}
To be consistent every else here

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/Client.scala [hunk 3]
```diff
@@ -292,22 +290,34 @@ private[spark] object Client {
...
       mainAppResource: String,
       mainClass: String,
       appArgs: Array[String]): Unit = {
+    val isPython = mainAppResource.endsWith(".py")
+    val pythonResource: Option[PythonSubmissionResourcesImpl] =
+      if (isPython) {
+        Option(new PythonSubmissionResourcesImpl(mainAppResource, appArgs))
+      } else None
+    // Since you might need jars for SQL UDFs in PySpark
+    def sparkJarFilter() : Seq[String] =
```
Comment (tnachen) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/Client.scala]: Style: sparkJarFilter(): Seq[String] =

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/ContainerLocalizedFilesResolver.scala [hunk 4]
```diff
@@ -24,13 +24,18 @@ private[spark] trait ContainerLocalizedFilesResolver {
...
 }
 
 private[spark] class ContainerLocalizedFilesResolverImpl(
     sparkJars: Seq[String],
     sparkFiles: Seq[String],
+    pySparkFiles: Seq[String],
+    primaryPyFile: String,
     jarsDownloadPath: String,
-    filesDownloadPath: String) extends ContainerLocalizedFilesResolver {
+    filesDownloadPath: String ) extends ContainerLocalizedFilesResolver {
```
Comment (tnachen) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/ContainerLocalizedFilesResolver.scala]: Remove extra space before )

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/DriverInitContainerComponentsProvider.scala [hunk 5]
```diff
@@ -32,13 +32,15 @@ import org.apache.spark.util.Utils
  */
 private[spark] trait DriverInitContainerComponentsProvider {
 
-  def provideContainerLocalizedFilesResolver(): ContainerLocalizedFilesResolver
+  def provideContainerLocalizedFilesResolver(
```
Comment (tnachen) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/DriverInitContainerComponentsProvider.scala:35]: Remove extra space after )

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/PythonSubmissionResources.scala [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+  def pySparkFiles: Array[String]
+  def arguments: Array[String]
+  def primaryPySparkResource(containerLocalizedFilesResolver: ContainerLocalizedFilesResolver)
+    : String
+  def driverPodWithPySparkEnvs(
+    driverPodFileMounter: DriverPodKubernetesFileMounter,
+    resolvedPrimaryPySparkResource: String,
+    resolvedPySparkFiles: String,
+    driverContainerName: String,
+    driverPodBuilder: PodBuilder) : Pod
```
Comment (tnachen) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/PythonSubmissionResources.scala]: Let's make the style consistent with the : spacing.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/PythonSubmissionResources.scala [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+  override def arguments: Array[String] = {
+    pyFiles.toList match {
+      case Nil => appArgs
+      case a :: b => a match {
+        case _ if a == mainAppResource && b == Nil => appArgs
+        case _ => appArgs.drop(1)
+      }
+    }
+  }
+  override def primaryPySparkResource (
```
Comment (tnachen) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/PythonSubmissionResources.scala]: As well with ( spacing

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/PythonSubmissionResourcesSuite.scala [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+import org.apache.spark.{SSLOptions, SparkConf, SparkFunSuite}
+import org.apache.spark.deploy.kubernetes.config._
+
+import scala.collection.JavaConverters._
+import io.fabric8.kubernetes.api.model.{ContainerBuilder, Pod, PodBuilder}
+import org.mockito.{Mock, MockitoAnnotations}
+import org.mockito.Mockito.when
+import org.scalatest.BeforeAndAfter
+
+
```
Comment (tnachen) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/PythonSubmissionResourcesSuite.scala]: Remove extra spaces

**Comment by tnachen** [comment]
Besides style problems, nothing jumps out to me as being wrong. Since it's tested already, I think it LGTM.

**Comment by ifilonenko** [comment]
@foxish waiting on your OK to merge

**Comment by foxish** [comment]
LGTM. Feel free to merge. Thanks!

**Comment by erikerlandson** [comment]
Likewise, LGTM

File: docs/running-on-kubernetes.md [hunk 9]
```diff
@@ -180,6 +180,32 @@ The above mechanism using `kubectl proxy` can be used when we have authenticatio
...
+      --conf spark.executor.memory=500m \
+      --conf spark.driver.memory=1G \
+      --conf spark.driver.cores=1 \
+      --conf spark.executor.cores=1 \
+      --conf spark.executor.instances=1 \
+      --conf spark.app.name=spark-pi \
+      --conf spark.kubernetes.driver.docker.image=spark-driver-py:latest \
+      --conf spark.kubernetes.executor.docker.image=spark-executor-py:latest \
+      --conf spark.kubernetes.initcontainer.docker.image=spark-init:latest \
+      --py-files local:///opt/spark/examples/src/main/python/sort.py \
```
Comment (holdenk) [inline docs/running-on-kubernetes.md:205]: I know this is unrelated, but so happy to have local support :)

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/ContainerLocalizedFilesResolver.scala [hunk 10]
```diff
@@ -24,14 +24,19 @@ private[spark] trait ContainerLocalizedFilesResolver {
...
   def resolveSubmittedSparkFiles(): Seq[String]
+  def resolveSubmittedPySparkFiles(): Seq[String]
+  def resolvePrimaryResourceFile(): String
 }
 
 private[spark] class ContainerLocalizedFilesResolverImpl(
     sparkJars: Seq[String],
     sparkFiles: Seq[String],
+    pySparkFiles: Seq[String],
+    primaryPyFile: String,
```
Comment (holdenk) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/ContainerLocalizedFilesResolver.scala:35]: Should this be an option?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/DriverInitContainerComponentsProvider.scala [hunk 11]
```diff
@@ -201,13 +206,15 @@ private[spark] class DriverInitContainerComponentsProviderImpl(
...
+  override def provideDriverPodFileMounter(): DriverPodKubernetesFileMounter = {
+    new DriverPodKubernetesFileMounterImpl()
+  }
   override def provideInitContainerBundle(
       maybeSubmittedResourceIds: Option[SubmittedResourceIds],
       uris: Iterable[String]): Option[InitContainerBundle] = {
-    val containerLocalizedFilesResolver = provideContainerLocalizedFilesResolver()
-    // Bypass init-containers if `spark.jars` and `spark.files` is empty or only has `local://` URIs
-    if (KubernetesFileUtils.getNonContainerLocalFiles(uris).nonEmpty) {
+    // Bypass init-containers if `spark.jars` and `spark.files` and '--py-rilfes'
```
Comment (holdenk) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/DriverInitContainerComponentsProvider.scala:215]: I know this is pre-existing but why?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/DriverInitContainerComponentsProvider.scala [hunk 12]
```diff
@@ -201,13 +206,15 @@ private[spark] class DriverInitContainerComponentsProviderImpl(
...
+    // is empty or only has `local://` URIs
+    if ((KubernetesFileUtils.getNonContainerLocalFiles(uris) ++ pySparkSubmitted).nonEmpty) {
```
Comment (holdenk) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/DriverInitContainerComponentsProvider.scala:217]: So right now pySparkSubmitted is defined as `val pySparkSubmitted = KubernetesFileUtils.getOnlySubmitterLocalFiles(pySparkFiles)` and that seems like not what is desired based on the previous comment and also the logic used for the non-container local files before hand.

File: resource-managers/kubernetes/docker-minimal-bundle/src/main/docker/driver-py/Dockerfile [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+RUN apk add --no-cache python && \
+    python -m ensurepip && \
+    rm -r /usr/lib/python*/ensurepip && \
+    pip install --upgrade pip setuptools && \
+    rm -r /root/.cache
+# UNCOMMENT THE FOLLOWING TO START PIP INSTALLING PYTHON PACKAGES
+# RUN apk add --update alpine-sdk python-dev
+# RUN pip install numpy
+
+ENV PYTHON_VERSION 2.7.13
```
Comment (holdenk) [inline resource-managers/kubernetes/docker-minimal-bundle/src/main/docker/driver-py/Dockerfile:36]: This seems like a weird version to set Python to? Do we expect people to roll their own Py3 containers?

File: resource-managers/kubernetes/docker-minimal-bundle/src/main/docker/executor-py/Dockerfile [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+    pip install --upgrade pip setuptools && \
+    rm -r /root/.cache
+# UNCOMMENT THE FOLLOWING TO START PIP INSTALLING PYTHON PACKAGES
+# RUN apk add --update alpine-sdk python-dev
+# RUN pip install numpy
+
+ENV PYTHON_VERSION 2.7.13
+ENV PYSPARK_PYTHON python
+ENV PYSPARK_DRIVER_PYTHON python
+ENV PYTHONPATH ${SPARK_HOME}/python/:${SPARK_HOME}/python/lib/py4j-0.10.4-src.zip:${PYTHONPATH}
```
Comment (holdenk) [inline resource-managers/kubernetes/docker-minimal-bundle/src/main/docker/executor-py/Dockerfile:39]: Could we source this from one of the existing pyspark scripts so it doesn't need to be updated?

If not, and we need this, we should add a comment about needing to update this when we update elsehwere. Also spark master is now @ 0.10.6.

**Review by holdenk — COMMENTED** [review]
I know this is merged, but I figured I'd leave some comments/questions if that works for folks?

**Pull request merged**